spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -85,48 +109,148 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
 					}
-					StreamListenerMessageHandler handler = new StreamListenerMessageHandler(invocableHandlerMethod);
-					handler.setApplicationContext(applicationContext);
-					handler.setChannelResolver(binderAwareChannelResolver);
-					if (!StringUtils.isEmpty(defaultOutputChannel)) {
-						handler.setOutputChannelName(defaultOutputChannel);
-					}
-					handler.afterPropertiesSet();
-					channel.subscribe(handler);
 				}
 			}
 		});
 		return bean;
 	}
 
+	private void invokeSetupMethodOnListenedChannel(Method method, Object bean) {
+		Object[] arguments = new Object[method.getParameterTypes().length];
+		for (int parameterIndex = 0; parameterIndex < arguments.length; parameterIndex++) {
+			MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+			Class<?> parameterType = methodParameter.getParameterType();
+			Input input = methodParameter.getParameterAnnotation(Input.class);
+			if (input != null) {
+				Object targetBean = this.applicationContext.getBean(input.value());
+				if (parameterType.isAssignableFrom(targetBean.getClass())) {
+					arguments[parameterIndex] = targetBean;
+				}
+				else {
+					for (StreamListenerArgumentAdapter<?, Object> streamListenerArgumentAdapter :
+							this.streamListenerArgumentAdapters) {
+						if (streamListenerArgumentAdapter.supports(targetBean.getClass(), methodParameter)) {
+							arguments[parameterIndex] = streamListenerArgumentAdapter.adapt(targetBean,
+									methodParameter);
+						}
+					}
+				}
+				if (arguments[parameterIndex] == null) {
+					throw new BeanInitializationException(
+							"Cannot convert argument " + parameterIndex + " of " + method + "from " + targetBean
+									.getClass().toString() + " to " + parameterType.toString());
+				}
+			}
+			else {
+				Output output = methodParameter.getParameterAnnotation(Output.class);
+				Object targetBean = this.applicationContext.getBean(output.value());
+				if (parameterType.isAssignableFrom(targetBean.getClass())) {
+					arguments[parameterIndex] = targetBean;
+				}
+				else {
+					for (StreamListenerArgumentAdapter<?, Object> streamListenerArgumentAdapter :
+							this.streamListenerArgumentAdapters) {
+						if (streamListenerArgumentAdapter.supports(targetBean.getClass(), methodParameter)) {
+							arguments[parameterIndex] = streamListenerArgumentAdapter.adapt(targetBean,
+									methodParameter);
+						}
+					}
+				}
+				if (arguments[parameterIndex] == null) {
+					throw new BeanInitializationException(
+							"Cannot convert argument " + parameterIndex + " of " + method + "from " + targetBean
+									.getClass().toString() + " to " + parameterType.toString());
+				}
+			}
+		}
+		try {
+			if (method.getReturnType().equals(Void.TYPE)) {
+				method.invoke(bean, arguments);
+			}
+			else {
+				Object result = method.invoke(bean, arguments);
+				Output output = AnnotationUtils.getAnnotation(method, Output.class);
+				if (output == null) {
```

Comment:
since this is already wrapped as a `BeanInitializationException` in the `catch` block below, this could just be an `Assert.notNull`